File: frontend/src/types/ranking.ts
export interface CropSummary {
  id: number;
  crop_name: string;
  scientific_name: string | null;
}

export interface FieldExplanation {
  short_explanation: string;
  detailed_explanation: string;
  strengths: string[];
  weaknesses: string[];
  risks: string[];
}

export interface ScoreComponentRead {
  key: string;
  label: string;
  weight: number;
  awarded_points: number;
  max_points: number;
  status:
    | "ideal"
    | "acceptable"
    | "limited"
    | "unconstrained"
    | "missing"
    | "blocked";
  reasons: string[];
}

export interface ScoreBlockerRead {
  code: string;
  dimension: string;
  message: string;
}

export interface RankedFieldRecommendation {
  rank: number;
  field_id: number;
  field_name: string;
  total_score: number;
  breakdown: Record<string, ScoreComponentRead>;
  blockers: ScoreBlockerRead[];
  reasons: string[];
  explanation: FieldExplanation;
}

export interface RankFieldsResponse {
  crop: CropSummary;
  total_fields_evaluated: number;
  ranked_results: RankedFieldRecommendation[];
}

export interface CropProfileRead extends CropSummary {
  ideal_ph_min: number;
  ideal_ph_max: number;
  tolerable_ph_min: number;
  tolerable_ph_max: number;
  water_requirement_level: string;
  drainage_requirement: string;
  frost_sensitivity: string;
  heat_sensitivity: string;
  salinity_tolerance: string | null;
  rooting_depth_cm: number | null;
  slope_tolerance: number | null;
  organic_matter_preference: string | null;
  notes: string | null;
  created_at: string;
  updated_at: string;
}

File: app/schemas/ranking.py
"""Schemas for the field ranking API."""

from __future__ import annotations

from pydantic import BaseModel, ConfigDict, PositiveInt

from app.engines.scoring_types import ScoreStatus
from app.schemas.explanation import FieldExplanation


class RankFieldsRequest(BaseModel):
    """Request payload for ranking fields against a selected crop."""

    model_config = ConfigDict(extra="forbid")

    crop_id: PositiveInt
    top_n: int | None = 5
    field_ids: list[PositiveInt] | None = None


class CropSummary(BaseModel):
    """Compact crop profile summary returned alongside ranking results."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    crop_name: str
    scientific_name: str | None


class ScoreComponentRead(BaseModel):
    """Serialized scoring component included in ranking responses."""

    model_config = ConfigDict(from_attributes=True)

    key: str
    label: str
    weight: float
    awarded_points: float
    max_points: float
    status: ScoreStatus
    reasons: list[str]


class ScoreBlockerRead(BaseModel):
    """Serialized hard blocker included in ranking responses."""

    model_config = ConfigDict(from_attributes=True)

    code: str
    dimension: str
    message: str


class RankedFieldRecommendation(BaseModel):
    """Single ranked recommendation entry for a field."""

    rank: int
    field_id: int
    field_name: str
    total_score: float
    economic_score: float
    estimated_profit: float | None
    ranking_score: float
    breakdown: dict[str, ScoreComponentRead]
    blockers: list[ScoreBlockerRead]
    reasons: list[str]
    explanation: FieldExplanation


class RankFieldsResponse(BaseModel):
    """Response payload returned by the field ranking endpoint."""

    crop: CropSummary
    total_fields_evaluated: int
    ranked_results: list[RankedFieldRecommendation]
